### PR TITLE
Add support for array arguments in vector mode

### PR DESCRIPTION
**Pull request by vaithak**

### CODE REVIEW

File: demos/VectorForwardMode.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// A typical invocation would be:
+// ../../../../obj/Debug+Asserts/bin/clang++  -Xclang -add-plugin -Xclang clad \
+// -Xclang -load -Xclang ../../../../obj/Debug+Asserts/lib/libclad.dylib     \
+// -I../include/ -x c++ -std=c++11 VectorForwardMode.cpp
+
+// Necessary for clad to work include
+#include "clad/Differentiator/Differentiator.h"
+
+// A function for weighted sum of array elements.
+double weighted_sum(double* arr, double* weights, int n) {
```
[the remark this answers is not in the document]
Comment (vaithak) [reply demos/VectorForwardMode.cpp:21]: Removed `demos` from clang-tidy.

File: lib/Differentiator/VectorForwardModeVisitor.cpp [hunk 2]
```diff
@@ -91,23 +97,83 @@ VectorForwardModeVisitor::DeriveVectorMode(const FunctionDecl* FD,
...
+      // Current offset for independent variable.
+      Expr* offsetExpr = arrayIndVarCountExpr;
+      Expr* nonArrayIndVarCountExpr = ConstantFolder::synthesizeLiteral(
+          m_Context.UnsignedLongTy, m_Context, nonArrayIndVarCount);
+      if (!offsetExpr) {
+        offsetExpr = nonArrayIndVarCountExpr;
+      } else if (nonArrayIndVarCount != 0) {
+        offsetExpr = BuildOp(BinaryOperatorKind::BO_Add, offsetExpr,
+                             nonArrayIndVarCountExpr);
+      }
```
Comment (vgvassilev) [inline lib/Differentiator/VectorForwardModeVisitor.cpp:139]: We should omit the `{` and `}`. Did clang-format add them?

File: lib/Differentiator/VectorForwardModeVisitor.cpp [hunk 3]
```diff
@@ -164,11 +236,12 @@ clang::FunctionDecl* VectorForwardModeVisitor::CreateVectorModeOverload() {
   }
 
   // instantiate output parameter type as void*
-  QualType outputParamType = m_Context.getPointerType(m_Context.VoidTy);
+  QualType outputParamType = GetCladArrayRefOfType(m_Context.VoidTy);
 
   // Push param types for derived params.
-  for (std::size_t i = 0; i < m_Function->getNumParams(); ++i)
+  for (std::size_t i = 0; i < m_Function->getNumParams(); ++i) {
```
Comment (vgvassilev) [inline lib/Differentiator/VectorForwardModeVisitor.cpp:242]: It seems we should fix clang-format's settings...

File: lib/Differentiator/VisitorBase.cpp [hunk 4]
```diff
@@ -611,6 +611,25 @@
     return InstantiateTemplate(GetCladArrayDecl(), {T});
   }
 
+  TemplateDecl* VisitorBase::GetCladMatrixDecl() {
+    static TemplateDecl* Result = nullptr;
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/VisitorBase.cpp:615]: I am not sure I understand that diagnostics. Does it mean to add a `const`?

**Review by vgvassilev — COMMENTED** [review]
This looks pretty good!

File: lib/Differentiator/VectorForwardModeVisitor.cpp (hunk 3 above)
Comment (vaithak) [reply lib/Differentiator/VectorForwardModeVisitor.cpp]: added `RemoveBracesLLVM` flag in clang-format, but this is not very reliable as stated here: https://clang.llvm.org/docs/ClangFormatStyleOptions.html

File: lib/Differentiator/VectorForwardModeVisitor.cpp (hunk 2 above)
Comment (vgvassilev) [reply lib/Differentiator/VectorForwardModeVisitor.cpp:139]: Probably. It's just a lexer in the end.

File: lib/Differentiator/VisitorBase.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/VisitorBase.cpp:615]: Not sure about this, I don't get it either.

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**